### PR TITLE
Add support for lambda permission import

### PR DESCRIPTION
**Pull request by chaspy**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/9357

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_lambda_permission: Support resource import
```

Output from acceptance testing:

```
make testacc TESTARGS='-run=TestAccAWSCloudWatchEventTarget_basic' 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSCloudWatchEventTarget_basic -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSCloudWatchEventTarget_basic
=== PAUSE TestAccAWSCloudWatchEventTarget_basic
=== CONT  TestAccAWSCloudWatchEventTarget_basic
--- PASS: TestAccAWSCloudWatchEventTarget_basic (58.64s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	58.721s
```

### CODE REVIEW

**Comment by chaspy** [comment]
@ryndaniels Please review 🙏

File: aws/resource_aws_lambda_permission_test.go [hunk 1]
```diff
@@ -188,6 +189,12 @@ func TestAccAWSLambdaPermission_basic(t *testing.T) {
 					resource.TestCheckResourceAttr("aws_lambda_permission.allow_cloudwatch", "event_source_token", "test-event-source-token"),
 				),
 			},
+			{
+				ResourceName:      "aws_lambda_permission.allow_cloudwatch",
+				ImportState:       true,
+				ImportStateId:     fmt.Sprintf("%s/%s", funcName, statementId),
+				ImportStateVerify: true,
+			},
```
Comment (ryndaniels) [inline aws/resource_aws_lambda_permission_test.go]: Hi @chaspy! Could you please add blocks like this to the rest of the acceptance tests in here, and re-run the acceptance tests once you've done that? Thanks!

File: website/docs/r/lambda_permission.html.markdown [hunk 2]
```diff
@@ -151,3 +151,11 @@ resource "aws_lambda_permission" "lambda_permission" {
  * `statement_id_prefix` - (Optional) A statement identifier prefix. Terraform will generate a unique suffix. Conflicts with `statement_id`.
 
 [1]: https://developer.amazon.com/docs/custom-skills/host-a-custom-skill-as-an-aws-lambda-function.html#use-aws-cli
+
+## Import
+
+Lambda Function Aliases can be imported using the function_name/statement_id, e.g.
```
Comment (ryndaniels) [inline website/docs/r/lambda_permission.html.markdown]: I think this should read something like: 
```suggestion
Lambda permission statements can be imported using the function_name/statement_id, e.g.
```
Could you double-check that please?

**Comment by chaspy** [comment]
```
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_basic'
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_StatementId_Duplicate'
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_withRawFunctionNamed'
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_withStatementIdPrefix'
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_withQualifier'
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_multiplePerms'
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_withS3'
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_withSNS'
make testacc TESTARGS='-run=TestAccAWSLambdaPermission_withIAMRole'
```

File: website/docs/r/lambda_permission.html.markdown (hunk 2 above)
Comment (chaspy) [reply website/docs/r/lambda_permission.html.markdown]: Addressed f925fb2

**Comment by chaspy** [comment]
```
$ make testacc TESTARGS='-run=TestAccAWSCloudWatchEventTarget_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSCloudWatchEventTarget_basic -timeout 120m

=== RUN   TestAccAWSCloudWatchEventTarget_basic
=== PAUSE TestAccAWSCloudWatchEventTarget_basic
=== CONT  TestAccAWSCloudWatchEventTarget_basic
--- PASS: TestAccAWSCloudWatchEventTarget_basic (67.61s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	67.690s

$ make testacc TESTARGS='-run=TestAccAWSLambdaPermission_withRawFunctionName' 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSLambdaPermission_withRawFunctionName -timeout 120m

=== RUN   TestAccAWSLambdaPermission_withRawFunctionName
=== PAUSE TestAccAWSLambdaPermission_withRawFunctionName
=== CONT  TestAccAWSLambdaPermission_withRawFunctionName
--- FAIL: TestAccAWSLambdaPermission_withRawFunctionName (112.13s)
    testing.go:568: Step 1 error: Cannot import non-existent remote object: While attempting to import an existing object to aws_lambda_permission.allow_cloudwatch, the provider detected that no object exists with the given id. Only pre-existing objects can be imported; check that the id is correct and that it is associated with the provider's configured region or endpoint, or use "terraform apply" to create a new remote object for this resource.
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	112.194s
```

😥

**Comment by ryndaniels** [comment]
Hi @chaspy! Thanks for the quick turnaround on those changes! I took a look at the new test failures you were seeing and I think I've found what's going on. It looks like there's a few mismatches in `statementId` and `ResourceName` in the different test blocks that are causing the failures we're seeing. 

One way that we could simplify the tests and fix those would be to create a `ImportStateIdFunc` function, the same way you did in #9431. I was working on testing that out locally, and realized that we'll need a bit more logic to handle `TestAccAWSLambdaPermission_withQualifier` and `TestAccAWSLambdaPermission_withStatementIdPrefix` correctly. I'm going to go ahead and continue working on getting those finished, and I'll keep this PR updated as I do. 🙂

**Comment by ryndaniels** [comment]
Latest acceptance test output:

```
--- PASS: TestLambdaPermissionGetFunctionNameFromLambdaArn_valid (0.00s)
--- PASS: TestLambdaPermissionUnmarshalling (0.00s)
--- PASS: TestLambdaPermissionGetQualifierFromLambdaAliasOrVersionArn_invalid (0.00s)
--- PASS: TestLambdaPermissionGetFunctionNameFromLambdaArn_invalid (0.00s)
--- PASS: TestLambdaPermissionGetQualifierFromLambdaAliasOrVersionArn_alias (0.00s)
--- PASS: TestLambdaPermissionGetQualifierFromLambdaAliasOrVersionArn_version (0.00s)
--- PASS: TestLambdaPermissionGetQualifierFromLambdaAliasOrVersionArn_govcloud (0.00s)
--- PASS: TestLambdaPermissionGetFunctionNameFromGovCloudLambdaArn (0.00s)
--- PASS: TestAccAWSLambdaPermission_withS3 (27.53s)
--- PASS: TestAccAWSLambdaPermission_withIAMRole (27.65s)
--- PASS: TestAccAWSLambdaPermission_withQualifier (27.97s)
--- PASS: TestAccAWSLambdaPermission_withRawFunctionName (28.06s)
--- PASS: TestAccAWSLambdaPermission_basic (27.91s)
--- PASS: TestAccAWSLambdaPermission_withStatementIdPrefix (28.22s)
--- PASS: TestAccAWSLambdaPermission_withSNS (28.23s)
--- PASS: TestAccAWSLambdaPermission_multiplePerms (34.30s)
--- PASS: TestAccAWSLambdaPermission_StatementId_Duplicate (83.63s)
```

File: aws/resource_aws_lambda_permission.go [hunk 3]
```diff
@@ -400,6 +403,41 @@ func getFunctionNameFromLambdaArn(arn string) (string, error) {
...
+	input := &lambda.GetFunctionInput{FunctionName: &functionName}
+
+	var qualifier string
+	fnParts := strings.Split(functionName, ":")
+	if len(fnParts) == 2 {
+		functionName = fnParts[0]
+		qualifier = fnParts[1]
+		input.Qualifier = &qualifier
+	}
+	statement := idParts[1]
```
Comment (ryndaniels) [inline aws/resource_aws_lambda_permission.go]:
```suggestion
	statementId := idParts[1]
```

File: aws/resource_aws_lambda_permission.go [hunk 4]
```diff
@@ -400,6 +403,41 @@ func getFunctionNameFromLambdaArn(arn string) (string, error) {
...
+	log.Printf("[DEBUG] Importing Lambda Permission %s for function name %s", statement, functionName)
```
Comment (ryndaniels) [inline aws/resource_aws_lambda_permission.go]:
```suggestion
	log.Printf("[DEBUG] Importing Lambda Permission %s for function name %s", statementId, functionName)
```

File: aws/resource_aws_lambda_permission.go [hunk 5]
```diff
@@ -400,6 +403,41 @@ func getFunctionNameFromLambdaArn(arn string) (string, error) {
...
+
+	conn := meta.(*AWSClient).lambdaconn
+	getFunctionOutput, err := conn.GetFunction(input)
+	if err != nil {
+		return nil, err
+	}
+
+	d.Set("function_name", getFunctionOutput.Configuration.FunctionArn)
+	d.Set("statement_id", statement)
```
Comment (ryndaniels) [inline aws/resource_aws_lambda_permission.go]:
```suggestion
	d.Set("statement_id", statementId)
```

File: aws/resource_aws_lambda_permission.go [hunk 6]
```diff
@@ -400,6 +403,41 @@ func getFunctionNameFromLambdaArn(arn string) (string, error) {
...
+	if qualifier != "" {
+		d.Set("qualifier", qualifier)
+	}
+	d.SetId(statement)
```
Comment (ryndaniels) [inline aws/resource_aws_lambda_permission.go]:
```suggestion
	d.SetId(statementId)
```

File: aws/resource_aws_lambda_permission.go [hunk 7]
```diff
@@ -400,6 +403,41 @@ func getFunctionNameFromLambdaArn(arn string) (string, error) {
 	return matches[5], nil
 }
 
+func resourceAwsLambdaPermissionImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	idParts := strings.Split(d.Id(), "/")
+	if len(idParts) != 2 || idParts[0] == "" || idParts[1] == "" {
+		return nil, fmt.Errorf("Unexpected format of ID (%q), expected FUNCTION_NAME/STATEMENT_ID", d.Id())
```
Comment (ryndaniels) [inline aws/resource_aws_lambda_permission.go]:
```suggestion
		return nil, fmt.Errorf("Unexpected format of ID (%q), expected FUNCTION_NAME/STATEMENT_ID or FUNCTION_NAME:QUALIFIER/STATEMENT_ID", d.Id())
```

File: aws/resource_aws_lambda_permission_test.go [hunk 8]
```diff
@@ -568,6 +624,20 @@ func lambdaPermissionExists(rs *terraform.ResourceState, conn *lambda.Lambda) (*
...
 
+func testAccAWSCLambdaPermissionImportStateIdFunc(resourceName string) resource.ImportStateIdFunc {
+	return func(s *terraform.State) (string, error) {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return "", fmt.Errorf("Not found: %s", resourceName)
+		}
+
+		if v, ok := rs.Primary.Attributes["qualifier"]; ok && v != "" {
+			return fmt.Sprintf("%s:%s/%s", rs.Primary.Attributes["function_name"], *aws.String(v), rs.Primary.Attributes["statement_id"]), nil
```
Comment (ryndaniels) [inline aws/resource_aws_lambda_permission_test.go]:
```suggestion
			return fmt.Sprintf("%s:%s/%s", rs.Primary.Attributes["function_name"], v, rs.Primary.Attributes["statement_id"]), nil
```

File: website/docs/r/lambda_permission.html.markdown [hunk 9]
```diff
@@ -151,3 +151,11 @@ resource "aws_lambda_permission" "lambda_permission" {
  * `statement_id_prefix` - (Optional) A statement identifier prefix. Terraform will generate a unique suffix. Conflicts with `statement_id`.
 
 [1]: https://developer.amazon.com/docs/custom-skills/host-a-custom-skill-as-an-aws-lambda-function.html#use-aws-cli
+
+## Import
+
+Lambda permission statements can be imported using the function_name/statement_id, e.g.
```
Comment (ryndaniels) [inline website/docs/r/lambda_permission.html.markdown]:
```suggestion
Lambda permission statements can be imported using function_name/statement_id, with an optional qualifier, e.g.
```

File: website/docs/r/lambda_permission.html.markdown [hunk 10]
```diff
@@ -151,3 +151,11 @@ resource "aws_lambda_permission" "lambda_permission" {
...
+
+```
+$ terraform import aws_lambda_function_permission.test_lambda_permission my_test_lambda_function/AllowExecutionFromCloudWatch
```
Comment (ryndaniels) [inline website/docs/r/lambda_permission.html.markdown]:
```suggestion
$ terraform import aws_lambda_function_permission.test_lambda_permission my_test_lambda_function/AllowExecutionFromCloudWatch

$ terraform import aws_lambda_function_permission.test_lambda_permission my_test_lambda_function:qualifier_name/AllowExecutionFromCloudWatch
```

**Pull request merged**